### DISCUSSION TITLE
Rework user level to be more contextually relevant

### DISCUSSION THREAD

**Discussion by willyosu** [ppy/osu#17124] [Ideas]
# Overview
Currently a user's level is determined purely based on their total score count from their profile, which arguably makes level a redundant and useless statistic. I've heard that there is [a goal to change level](https://github.com/ppy/osu/discussions/15710#discussioncomment-1676159) in the future, but not really any set plans about how to do so (at least none that I've heard in public). I'd like to just throw out a few ideas for improving level even if some of this may have been discussed in dev channels or some other place prior.

# Ideas

**1. Create a new statistic to replace total score as a basis for levels**

A similar (or even the same) formula to compute level can be used, but the idea is having a new value like "xp" to replace it. Certain activities can add to a user's "xp" total, such as gaining ranked score, setting a score on a new beatmap, and/or (most intuitively) each time a user unlocks a new medal. As a user does different activities like these their level goes up, much like it would in other games.

- All of these events should be easily trackable to not add any additional overhead to the system. For the examples given profiles already access the ranked score, total SS/S/A counts, and total medal count respectively. This would just be a matter of multiplying the numbers by a predetermined value and summing them.
- Events should be irreversible so that a user cannot randomly lose "xp". In every game the basic concept of experience and levels are a total which is continually rising, so we would need to sample stats which can only gain to achieve this. I am weary of including so-called "pp milestones" or similar events due to the fact that the user can overwrite the score with a lower value causing their "xp" and level to go down.

- Here are a few of the user statistics which have potential to be used towards the xp count:
    - medal count
    - highest combo
    - ranked score
    - total hits
    - total submitted scores *a.k.a. rank counts* (SS/S/A)
    - play time

**2. Revamp level formula to better suit players**

It may sound strange at first but this is an issue that is made very apparent from the current way level is computed, which is as follows:
```
score(n) = 5,000 / 3 * (4n^3 - 3n^2 - n) + 1.25 * 1.8^(n - 60) if n <= 100
score(n) = 26,931,190,827 + 99,999,999,999 * (n - 100)         if n > 100
```
The obvious hole being that this piecewise function simply stops being exponential at level 100 and from that point on a user just needs to keep gaining ~100B score per level. Beyond this lower levels are extremely easy to achieve, especially considering that playing an Easy or Normal difficulty **once** and getting ~2,000,000 score will instantly bring you to almost all the way to level 7. Reworking this function to be more user-friendly at the lower end and more challenging at the upper end could greatly improve how level works (although finding this may be harder than I believe). 

- The exponential curve should be less harsh. As it stands for beginner players I've talked to getting to about level 40 or 50 is very easy, whereas they find the 80s, 90s, and beyond to be quite a height. Gaining levels should feel rewarding and not something which is overly easy early on and overly difficult later.
- A level formula should be continuous rather than piecewise (or just not linear!).The last point is contrasted by very skilled players who amass hundreds of billions of total score relatively (and I do mean relatively) easily through longer and higher score rewarding maps. 
- Of course, any function given is dependent on the values which are attributed to the "xp" stat going forward, so I won't propose a specific formula for this. Attributes or events which can contribute should be determined before fitting values to each and setting a formula overall. Maybe a curve such as `x^( 1 / e )` or `ln( x + 1 )` could be very simple candidates as a starting point to fit this function if you gave it some other coefficients.

-----

There is lots of potential here to make user level much more relevant and useful given these ideas, or at least I think so. Constructing the new system/formula will definitely take some tweaking of values and fitting of curves for this.

I hope I've outlined some good ideas to carry over into the next iteration of level, and that others will be able to build upon and think of an even better system.

**Comment by peppy** [comment]
The original plan was to make level actually be required to play certain mods (at all? for ranking?) to give a sense of unlocking in the game. I think it would be cool to outline other things we could potentially make level based unlocks to increase the excitement and goals of new players.

**Reply by willyosu** [reply]
That seems really interesting and would definitely give level a clear use in the future. My only worry with it being not allowing players to play what they want, but I assume the mod unlocking would apply at such a low level that it would actually benefit the new players giving them a path to progress.

Apart from that if there aren't any detailed plans for how exactly level would be gained, I'm curious to know if you think some of the user statistics which I outlined would be good candidates for "giving xp" and if there may be any interest in implementing this soon.

**Reply by Givikap120** [reply]
This is good, but I feel like there should be option to unlock everything
Handholding system that will recommend maps and mods is good as a tutorial for beginners
But this should be optional if player really want to do everything that they want

**Comment by respektive** [comment]
I like the idea of going more towards a "xp" system using multiple stats. Medal count is one of things i actually never thought about but it would probably make for a pretty good base stat to expand on instead of total score.  
One small thing i see with this tho is that medal counts are across all rulesets which would mean level would also display for all rulesets then and isn't split anymore. If that's a direction we wanna go anyway it's no problem.  
Otherwise if we want to keep the level stat separated per ruleset we would have to stick to other stats as a base that are tracked per ruleset and use the medal count as an expansion on top of it.  
Overall you bring up some really good points though!

**Reply by SensouShin** [reply]
Or we can use both idea, keeping separate level for each ruleset and add a new Overall level which is contains all XP, so we can see how much a ruleset contributed to the Overall level

**Reply by iSimpFor** [reply]
i agree with the overall level since it shows a user with excellent experience in the game compared to another user 

each medal should give a different amount of xp

i think there should be a level system for mapping related stuff

**Reply by willyosu** [reply]
If we do want to split up per ruleset, then I think storing of an "xp" number would have to be the way to go, much like how total score operates today. That way you could add to the xp total for a certain ruleset when that ruleset-specific medal is unlocked (and also potentially have different xp values for different medals!).

**Reply by respektive** [reply]
Different xp for different medals would be pretty cool.  
For mapping related xp, using the already existing kudosu values could work, aswell as using maybe ranked maps count and maps nominated for BNs.

**Reply by FunnyFortniteFan** [reply]
bumping this as it would be really cool to see level reworked.

**Comment by WitherFlower** [comment]
peppy's recent twitter post about making a new metric for related to dedication sparked my interest, so I would like to share a proposal for that metric, as well as a proposal to revamp the level system to use more statistics.

# New Statistic (dedication points)

The calculation is the following : 
```cs
ClearValue = SSCount * 6 + FCCount * 2 + ClearCount
ScoreValue = RankedScore / 1_000_000_000
DedicationPoints = Sqrt(ClearValue * ScoreValue)
```
(The above assumes that an SS will contribute to all 3 counts, meaning it actually increases clear value by 9, and likewise, an FC increases clear value by 3.)

This gives the following leaderboard (made by @EngineerMark) : 
![image](https://github.com/ppy/osu/assets/52672543/42ca2a47-e729-4c8e-816d-428161113882)

In my original twitter proposal, I inculded a bonus for medals to the total, which could be as simple as `MedalCount*20`, but I think it's better if medals have different rewards based on difficulty. Medals also have the issue that most of them aren't ruleset-specific, so that would require more consideration as well.

# Level system revamp proposal

As proposed by OP, I included more stats to contribute to level.
Each stat contributes a certain amount of XP : 
| Stat | XP gain per unit |
| --- | --- |
| SS Count  | 100 |
| S Count | 50 |
| A Count | 25 |
| Ranked Score | 1/500,000 |
| Total Score | 1/500,000 |
| Medals (1) | 10000 |
| Performance (2) | 250 |
| Hours of playtime (3) | 200 |

The level-to-XP formula goes as follows : 

`Total XP Required = 5 * Level^3 + 120 * Level^2 + 105 * Level - 230`

This makes players reach level 100 at a similar point than the current system, while keeping levels going past that without feeling like the current system's wall. The highest level players with this formula sit at around Lvl150.

You can play around with the values using [this spreadsheet](https://docs.google.com/spreadsheets/d/1WndQ0ozEyVzdIXqirH-xg9hMx6omo7jNqSLfZ25lOow/edit?usp=sharing)

**Notes :**

(1) Ideally, medals would be separated by ruleset and give different XP values depending on their rarity / difficulty. It doesn't feel right for the 1 star pass and 10 star FC medals to give the same amount of XP.

(2) Performance is also a bit weird because PP reworks could cause levels to go down, but it can be removed if skill-based medals give good XP rewards instead. 

(3) Playtime is included instead of playcount because I don't see a valid reason why playing a 10 minute map should give you less XP than playing 10 1-minute maps.

**Reply by Walavouchey** [reply]
to be clear, you're proposing either the dedication points or the level rework, right?

imo pp shouldn't be considered for levels precisely because of the reason given, and it can potentially update very frequently. medals could definitely give manually assigned xp values

values will like any xp system have to be arbitrary, and i'd like to think a proposal like this to start off is in the correct direction

**Reply by Walavouchey** [reply]
cc @ppy/team-client as this is in the [future tasks list](https://github.com/ppy/osu/issues/27686)

**Reply by WitherFlower** [reply]
I am proposing that both systems get implemented.
I imagine dp would come with its own ranking, but not level (although it would be really cool as well). 

I also feel like they represent different things. dp should represent "dedication", level should represent "experience".

Level already kind of does represent "experience", but the idea of reworking it is to make it fun and relevant instead of just being a clone of Total Score.

**Reply by Walavouchey** [reply]
> I also feel like they represent different things. dp should represent "dedication", level should represent "experience".

but they take the same statistics... it's like having two different scoring systems (oh wait we do)

two new leaderboards aside you can still compete with levels in the same way as "dedication points", same way as total score, same way as medal counts (potentially with external leaderboards if there aren't official ones)

**Reply by WitherFlower** [reply]
If the goal is to add a new dynamic to the game, then whatever stat ends up being used needs a ranking. It's perfectly fine if only one of the two options (or another similar system) ends up being used as long as there is a ranking that appears on profiles.

On that topic, when that ranking is introduced, players should be able to set this as their "main ranking" in game, so it's what shows up on their player card. Or maybe both this and the pp ranking could be displayed side by side.

**Comment by WitherFlower** [comment]
I made a revision of my level rework proposal, removing the contribution from performance points, and adjusting the values to improve the balance and scaling.

You can see a ranking of players by this level metric [here](https://score.kirino.sh/leaderboard/stat/dedication_level/page/1/country/world), and the top 20 is on the image below.

![image](https://github.com/ppy/osu/assets/52672543/41ad47a6-98f5-4d34-8144-0a0d2311474b)

Here are the new values / level formula : 

| Stat | XP gain per unit |
| --- | --- |
| SS Count  | 200 |
| S Count | 100 |
| A Count | 50 |
| Ranked Score | 1/125,000 |
| Total Score | 1/250,000 |
| Medals (1) | 20,000 |
| Hours of playtime | 300 |

`Total XP Required = 5 * Level^3 + 80 * Level^2 + 225 * Level - 310`

You can play around with the values using [this spreadsheet](https://docs.google.com/spreadsheets/d/1WndQ0ozEyVzdIXqirH-xg9hMx6omo7jNqSLfZ25lOow/edit#gid=1263339561)

**Notes :**
(1) Medals should be given a different reward based on their level of difficulty. Consider this value of 20,000 to be what the average should be amongst all of them.

**Reply by senpai-haze-sama** [reply]
SS = 200 S = 100, it is double. in my opinion it shouldn't be so influential to play well because it's not a question of skill but of dedication, so I would opt to reduce the difference in points between SS, S and others. even if you say that more dedication is needed to do SS it is not a question directly linked to the result. if you make an attempt to do an SS and then an attempt to do an S. the dedication you put in is the same, this is a DP system not a PP system. and in any case the fact of being at a higher skill level still means that you play more so it is rewarded in any case

**Reply by WitherFlower** [reply]
Thanks for your feedback ! Here is my reasoning behind those values for SS ranks.
I think that most maps don't require "more skill" to achieve an SS on as they require "more retries". Because most maps in osu are really easy (59% of the entire ranked section falls below 4 stars, that's 65000 maps), getting an SS on all of those maps (barring the few weird map edge cases) doesn't require that much skill compared to the time investment and dedication needed. Taking a look at the ranking obtained with the current values, I think that it is **very** fair for Zalaria and Sleepteiner to rank in the top 10, given the amount of time and effort they have put into farming SS Ranks.

**Reply by senpai-haze-sama** [reply]
it remains a question of skill. and dedication translates into more playtime and more playcount so it is already rewarded in itself. rather, in addition to breaking down the difference between the various grades, maybe you should balance the rest of the factors a little better instead, for example by also adding the hit count to the algorithm. we need to think impartially in these cases, and not think about favoring some players over others

**Reply by WitherFlower** [reply]
Playtime is already factored into the formula. However, I think some skill has to be involved at the end of the day, or the metric will be uninteresting / unfun. Nobody would like to mindlessly farm playtime or hit counts for literal thousands of hours. That's why the current level system is bad and no one really cares about it.

**Reply by senpai-haze-sama** [reply]
I disagree with you. you're deliberately favoring one type of gameplay. instead we don't have to look at the players. but we must create a system that is impartial and mathematically consistent with the concept of "dedication"

**Reply by 6800cube** [reply]
This is really excited idea for me, a farmer who has been playing for 14 years. 
Thank you for all the efforts made in defining the level, and before implementation, we can first identify a few issues.
1. The ranking of players can be measured. I think an SS is worth at least 2-3 times compared to S, because they are really too difficult! That's why I don't have much SS. But for some maps, S may be similar to A, because of the presence of slider objects, many spectra without combos can also achieve S ranking (I think some high combo A ranking is more difficult than low combo S ranking).
2. If both the ranking score and total score are measured, there will be duplicate parts. For example, if you create a new spectrum, you will receive two different score bonuses. I think this is very reasonable, but the key is the ratio of the two scores and the relationship with the number of rankings. For me (a relatively average player), an S is roughly equal to a score of 10 million.
3. The difficulty and experience of medals, although I have many medals, I think some simple medals are really not worth it. I believe you can make them better.

**Reply by 6800cube** [reply]
Thank you for everything you have done to the level system. Looking forward to its implementation!

**Reply by lillip0p** [reply]
i like this system! i think each medal could have a custom xp amount

**Reply by Hiviexd** [reply]
> i like this system! i think each medal could have a custom xp amount

internally, we gave a lot of medals a rough "score" based on their general difficulty. said score is currently unused, so it may be a good way forward if we're weighing in medals in an XP system.

**Reply by WitherFlower** [reply]
Here is a demonstration of the system also working well with a Clan/Team system, courtesy of @EngineerMark
It could also tie into other metrics at clan-level, if they ever get added.

![image](https://github.com/user-attachments/assets/c0db6866-8293-449f-a40b-53a5764d1c08)

**Comment by AosUK** [comment]
What do people think about purely, and unweighted summing pp. 
It would be possible to mindlessly grind.
The nature of the pp system already considerers both the difficulty of the map and the players performance on it.
Anyone can sit down and play new maps (within their skill range) and progress, while more skilled players will progress faster, and have more maps available to them
Yes this is very simple, but only because all the heavy lifting for determining the worth of an individual score has already been done (to what degree of accuracy is another question) 

Potential issues I see:
This is up for debate as I don't see it as a huge problem, but others may - 20 50pp plays would be the same as 1 1000pp play.
I see this as fine because we already have the weighted system.
A possible solution could be using an *unweighted* sum of the top 2,000 pp plays of a user 

This also forgets about medals, or other things (example: interesting mod combinations) we want players to experience, they're essentially still going to pp farm just in a different way.

There are many more issues, I will leave articulating those to people more knowledgeable.
At first glance this would seem to be a good balance of the weighted pp system we have, and (lets say) a users play count, skill is rewarded but it isn't impossible to progress without it

**Reply by minisbett** [reply]
This would cause an unbalanced measurement of "dedication" as higher skilled players need to put much less dedication for the same result than less skilled players.
